### PR TITLE
OAuth from GitHub API

### PR DESCRIPTION
**Pull request by jzlotek**

### CODE REVIEW

File: components/routes/github-authentication.js [hunk 1]
```diff
@@ -1,23 +1,45 @@
...
 module.exports = function(webserver, controller) {
 
     debug('Configured /github/auth url');
     webserver.get('/github/auth', (req, res) => {
+        console.log('entered on /github/auth')
         let code = req.query.code
+        let access_token = req.query.access_token
+        let token_type = req.query.token_type
         if (code) {
+            console.log(code)
```
Comment (dgautsch) [inline components/routes/github-authentication.js]: let's use `debug('Bot Code: ', code);`

File: components/routes/github-authentication.js [hunk 2]
```diff
@@ -1,23 +1,45 @@
 var debug = require('debug')('botkit:incoming_webhooks');
 var axios = require('axios')
 
+const state = "fndsjgbfhusdabic457438hfb89b0qbd80n9u2n4qf80dxrwb0t"
```
Comment (dgautsch) [inline components/routes/github-authentication.js]: This should be stored in an ENV variable.

File: components/routes/github-authentication.js [hunk 3]
```diff
@@ -1,23 +1,45 @@
...
+            })
 
+        } else if (access_token && token_type) {
+            console.log(access_token, token_type)
         } else {
             console.log("resonse with no code")
         }
 
-        res.redirect('/')
+        // res.redirect('/')
```
Comment (dgautsch) [inline components/routes/github-authentication.js]: Let's remove this, unless we plan on bringing it back.

File: components/routes/github-authentication.js [hunk 4]
```diff
@@ -1,23 +1,45 @@
...
     })
     
     webserver.get('/github', (req, res) => {
-        return res.redirect(`https://github.com/login/oauth/authorize?scope=user:email,read:user,read:org&client_id=${GITHUB_CLIENT_ID}`)
+        console.log('entered on /github')
```
Comment (dgautsch) [inline components/routes/github-authentication.js]: use debug here.

File: components/routes/github-authentication.js [hunk 5]
```diff
@@ -1,23 +1,45 @@
...
+            console.log(code)
             // add OAuth code to db
+            // axios.post('https://github.com/login/oauth/access_token', {
+            //     client_id: `${process.env.GITHUB_CLIENT_ID}`,
+            //     client_secret: `${process.env.GITHUB_CLIENT_ID}`,
+            //     code: `${code}`,
+            //     redirect_uri: 'http://localhost:3000/github/auth',
+            //     state: state
+            // })
+            axios.post(`https://github.com/login/oauth/access_token?client_id=${process.env.GITHUB_CLIENT_ID}&client_secret=${process.env.GITHUB_CLIENT_SECRET}&code=${code}&state=${state}&redirect_uri=http://localhost:3000/github/auth`)
```
Comment (seriouslysean) [inline components/routes/github-authentication.js]: Convert these params to the axios object syntax:
```
axios.post(url, {
  key: process.env.VALUE,
  ...[etc],
}).then(...)
```

```suggestion
            axios.post('https://github.com/login/oauth/access_token', { 
                client_id: process.env.GITHUB_CLIENT_ID,
                client_secret: process.env.GITHUB_CLIENT_SECRET,
                code,
                state,
                redirect_uri: 'http://localhost:3000/github/auth'
            )
```

File: components/routes/github-authentication.js [hunk 6]
```diff
@@ -1,23 +1,45 @@
...
         }
 
-        res.redirect('/')
+        // res.redirect('/')
     })
     
     webserver.get('/github', (req, res) => {
-        return res.redirect(`https://github.com/login/oauth/authorize?scope=user:email,read:user,read:org&client_id=${GITHUB_CLIENT_ID}`)
+        console.log('entered on /github')
+        return res.redirect(`https://github.com/login/oauth/authorize?scope=user:email,read:user,read:org&client_id=${process.env.GITHUB_CLIENT_ID}&state=${state}`)
```
Comment (seriouslysean) [inline components/routes/github-authentication.js]: Convert to something like this:
```suggestion
        return res.redirect(
            'https://github.com/login/oauth/authorize' +
            '?scope=user:email,read:user,read:org' + 
            `&client_id=${process.env.GITHUB_CLIENT_ID}` + 
            `&state=${state}`
          );
```

Untested, but something a bit more readable here is the goal.

File: components/routes/github-authentication.js [hunk 7]
```diff
@@ -1,23 +1,45 @@
...
+            .then(result => {
+                console.log(result.data.split('&').forEach(e => {
+                    console.log(e.split('='))
+                }))
+            })
 
+        } else if (access_token && token_type) {
+            console.log(access_token, token_type)
         } else {
             console.log("resonse with no code")
```
Comment (seriouslysean) [inline components/routes/github-authentication.js]: Typo, should be `response with no code`:
```suggestion
             console.log("response with no code")
```

File: components/routes/github-authentication.js [hunk 8]
```diff
@@ -1,23 +1,45 @@
...
+const state = "fndsjgbfhusdabic457438hfb89b0qbd80n9u2n4qf80dxrwb0t"
+
 module.exports = function(webserver, controller) {
 
     debug('Configured /github/auth url');
     webserver.get('/github/auth', (req, res) => {
+        console.log('entered on /github/auth')
         let code = req.query.code
+        let access_token = req.query.access_token
+        let token_type = req.query.token_type
```
Comment (seriouslysean) [inline components/routes/github-authentication.js]: You can probably do something like this instead:
```
const { code, access_token, token_type } = req.query;
```

File: components/routes/github-authentication.js [hunk 9]
```diff
@@ -1,23 +1,51 @@
...
+                })
 
+            }).catch(err => debug(err))
+
+
+        } else if (access_token && token_type) {
+            debug(access_token, token_type)
         } else {
-            console.log("resonse with no code")
+            debug("response with no code")
```
Comment (seriouslysean) [inline components/routes/github-authentication.js]: Might want to make this an error since we expect a code?

**Review by dgautsch — APPROVED** [review]

**Pull request merged**